SoundSource Plugin API v10: Fallback providers

uklotzde opened the pull request:
A simple extension that allows to register multiple SoundSource providers for the same file extension. The providers are used in the order of their priority with which they were registered. SoundSourceProxy prints a debug log message that shows which providers have been registered at startup:

Debug [Main]: SoundSource providers for file extension "aac" 
Debug [Main]:   0 : "FFmpeg" 
Debug [Main]: SoundSource providers for file extension "aif" 
Debug [Main]:   1 : "libsndfile" 
Debug [Main]: SoundSource providers for file extension "aiff" 
Debug [Main]:   1 : "libsndfile" 
Debug [Main]: SoundSource providers for file extension "flac" 
Debug [Main]:   2 : "Xiph.org libFLAC" 
Debug [Main]:   1 : "libsndfile" 
Debug [Main]:   0 : "FFmpeg" 
Debug [Main]: SoundSource providers for file extension "it" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "m4a" 
Debug [Main]:   2 : "Nero FAAD2" @ "/home/uk/.mixxx/plugins/libsoundsourcem4a.so" 
Debug [Main]:   0 : "FFmpeg" 
Debug [Main]: SoundSource providers for file extension "med" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "mod" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "mp3" 
Debug [Main]:   2 : "MAD: MPEG Audio Decoder" 
Debug [Main]:   0 : "FFmpeg" 
Debug [Main]: SoundSource providers for file extension "mp4" 
Debug [Main]:   2 : "Nero FAAD2" @ "/home/uk/.mixxx/plugins/libsoundsourcem4a.so" 
Debug [Main]: SoundSource providers for file extension "ogg" 
Debug [Main]:   2 : "Xiph.org OggVorbis" 
Debug [Main]:   0 : "FFmpeg" 
Debug [Main]: SoundSource providers for file extension "okt" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "opus" 
Debug [Main]:   2 : "Xiph.org libopusfile" 
Debug [Main]: SoundSource providers for file extension "s3m" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "stm" 
Debug [Main]:   2 : "MODPlug" 
Debug [Main]: SoundSource providers for file extension "wav" 
Debug [Main]:   1 : "libsndfile" 
Debug [Main]: SoundSource providers for file extension "wv" 
Debug [Main]:   2 : "WavPack" @ "/home/uk/.mixxx/plugins/libsoundsourcewv.so" 
Debug [Main]: SoundSource providers for file extension "xm" 
Debug [Main]:   2 : "MODPlug" 

In this example 2 providers are availabe for decoding of .m4a files: "Nero FAAD2" with default priority 2 (provided as a plugin by an external library) and "FFmpeg" with the lowest priority 0 as a fallback.

SoundSourceProxy automatically switches to the next registered provider whenever creating a SoundSource or opening an AudioSource fails. In the following example the "Nero FAAD2" plugin fails to open an MP4 file with DASH content (ftyp="dash") and "FFmpeg" is used as a fallback:

Debug [AnalyserQueue 1]: Obtained SoundSource for  QUrl( "file:///home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a" )  from provider "Nero FAAD2"
Debug [Main]: WSpinny::slotCoverFound WSpinny(0x505cb30, name = "SpinnySingleton1")  "CoverInfo(NONE,GUESSED,,0,/home/uk/Music/Corrupt Tracks (Link)/faad_test_6n73tOSV6-M.m4a)" QSize(0, 0)
Warning [AnalyserQueue 1]: Failed to read number of samples from file: "file:///home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a"
Warning [AnalyserQueue 1]: Failed to open AudioSource for file "file:///home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a" with provider "Nero FAAD2"
Debug [AnalyserQueue 1]: Obtained SoundSource for  QUrl( "file:///home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a" )  from provider "FFmpeg"
Debug [AnalyserQueue 1]: New SoundSourceFFmpeg : "/home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a"
Debug [AnalyserQueue 1]: ffmpeg: NEW FFMPEG API using libavresample
Debug [AnalyserQueue 1]: Created sample rate converter for conversion of 44100 Hz format: fltp with: 2 (layout: 3 ) channels (BPS 4 )
Debug [AnalyserQueue 1]: To  44100  HZ format: flt with 2 (layout: 3 ) channels (BPS  4 )
Debug [AnalyserQueue 1]: ffmpeg: Samplerate:  44100 , Channels:  2
 Debug [AnalyserQueue 1]: Opened AudioSource for file "file:///home/uk/Music/Corrupt Tracks/faad_test_6n73tOSV6-M.m4a" with provider "FFmpeg"

uklotzde committed:
Give each SoundSourceProvider an inherent priority

uklotzde committed:
Retry to open Sound-/AudioSources with different providers after failure

uklotzde committed:
C++11: Replace foreach with range-based for loop

uklotzde committed:
Improve readability of code

uklotzde committed:
Change debug log message

uklotzde commented:
I'm aware that unit tests are still missing. The SoundSourceProviderRegistry is perfectly suited for unit testing.

uklotzde committed:
Delete obsolete functions

uklotzde committed:
Harmonize SoundSourceProxy with TIO branch to minimize merge conflicts

uklotzde committed:
Add unit tests for SoundSourceProviderRegistry

uklotzde committed:
C++11: Use strongly typed enum for provider priority

uklotzde committed:
Move Priority out of SoundSourceProvider

uklotzde committed:
Fix build error after merging master

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -248,51 +193,53 @@ void SoundSourceProxy::loadPlugins() {
+        qDebug() << "SoundSource providers for file extension" << supportedFileExtension;
+        const auto registrationsForFileExtension(
+                s_soundSourceProviders.getRegistrationsForFileExtension(
+                        supportedFileExtension));
```
Sorry, I am struggling with this notation. The return type of getRegistrationsForFileExtension is out of sight and
GitHub has no tooltip for it. You may argue that the real type does not add any value, because it is used below as it is a container. But below you use auto as well (which would be nice if the reader knows the content).

getRegistrationsForFileExtension returns SoundSourceProviderRegistrationList which is a typedef that saves 3 characters and also hides information. I would prefer to use here QList<SoundSourceProviderRegistration>  directly. 

```
        const QList<SoundSourceProviderRegistration> registrationsForFileExtension(
                s_soundSourceProviders.getRegistrationsForFileExtension(
                        supportedFileExtension));
        for (const auto& registration: registrationsForFileExtension) {
 ```

uklotzde replied:
The main purpose of the typedef is not to save any characters ;) I will remove it improves readability.

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -248,51 +193,53 @@ void SoundSourceProxy::loadPlugins() {
+    s_supportedFileNamePatterns.clear();
+    for (const auto& supportedFileExtension: supportedFileExtensions) {
+        s_supportedFileNamePatterns += QString("*.%1").arg(supportedFileExtension);
+    }
```
There is either a space missing in the format or additional spaces in the comment string.

uklotzde replied:
Ups. This bug is already present in master! I will fix it here and separately also there.

uklotzde replied:
And this code is correct, because the individual patterns are concatenated as a QStringList.

uklotzde replied:
I will fix the misleading comment.

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -308,31 +255,163 @@ bool SoundSourceProxy::isFileSupported(const QFileInfo& fileInfo) {
+        closeAudioSource();
+        m_pSoundSource.clear();
+    }
+    initSoundSource();
```
It this a recursive call?

uklotzde replied:
In fact an unnecessary recursion! I didn't notice, easy to fix.

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -308,31 +255,163 @@ bool SoundSourceProxy::isFileSupported(const QFileInfo& fileInfo) {
+          m_pSecurityToken(openSecurityToken(pTrack)),
+          m_soundSourceProviderRegistrations(findSoundSourceProviderRegistrations(m_url)),
+          m_soundSourceProviderRegistrationIndex(0) {
+    initSoundSource();
```
can we move this to openAudioSource?

daschuer replied:
This way we have a single function that is responsible to init a working SoundSource

uklotzde replied:
The SoundSource is needed for parsing metadata before the audio decoder is actually required. We should not merge this for performance reasons. Initializing the decoder just because we want to parse some metadata is too expensive.

As a consequence metadata and audio might be read by different SoundSources, if the SoundSource that was used to read the metadata fails to open the audio stream. Currently not really an issue, because metadata will almost always be read by the generic TagLib functions. The only solution is to decouple access to metadata and audio. But this would render the concept of a SoundSource that combines both MetadataSource and AudioSource interfaces useless.

daschuer commented on src/sources/soundsourceffmpeg.h:
```diff
@@ -88,6 +88,11 @@ class SoundSourceProviderFFmpeg: public SoundSourceProvider {
 
+    SoundSourceProviderPriority getPriorityHint() const override {
+        // FFmpeg will only be used as the last resort
+        return SoundSourceProviderPriority::LOWEST;
```
A single priority hint might fail. We may have a criss cross dependency. 
If a sound source has the best quality for a single file format but supports other formats in a weak way.
I am not sure if a priority hint for each file extension will be sufficient. Since it is sometimes only a extension from a container format that can carry any encoding type.

uklotzde replied:
We could add a parameter with the file extension as you mentioned below. This will be sufficient until Mixxx determines the file type by the actual content instead of only using the file extension.

daschuer replied:
This will work for me.

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,85 @@
+            // Priority comparison with <=: New registrations will be inserted
+            // before existing registrations with equal priority, but after
+            // existing registrations with higher priority.
+            if (i->getProviderPriority() <= registration.getProviderPriority()) {
```
yes, one priority for each file extension.

daschuer commented on src/soundsourceproxy.cpp:
```diff
@@ -219,24 +170,18 @@ void SoundSourceProxy::loadPlugins() {
-    foreach (QDir dir, pluginDirs) {
-        qDebug() << "Loading SoundSource plugins" << dir.path();
-        const QStringList files(dir.entryList(
+    for (const auto& pluginDir: pluginDirs) {
```
The plugins are registered with there priority hint. This requires that the priority hint is set cooperative.
Can we guarantee this or do we need a facility to set the priority from the Mixxx source.
Or should the priority be read from a config file? Any ideas?

uklotzde replied:
I think the cooperative approach is sufficient for the moment.

I've added functionality to deregister a provider for a specific file extension together with a function to (re-)register a provider for a specific file extensions with a custom priority. These 2 functions can be used to change the default priority of a provider for a specific file extension.

Modifying the provider registry at runtime might require additional locking, but all the required functionality will already be available.

uklotzde committed:
Delete typedef SoundSourceProviderRegistrationList (review comments)

uklotzde committed:
Replace auto with explicit type (review comments)

uklotzde committed:
Fix misleading comment (review comments)

uklotzde committed:
Add deregistration functions for SoundSource providers

uklotzde committed:
Remove recursion from nextSoundSourceProvider() (review comments)

uklotzde committed:
Refactoring and optimizations for C++11 (std::move)

uklotzde committed:
Add function to modify SoundSource provider priorities

uklotzde committed:
Rename class member

uklotzde committed:
Parameterize priority hints by file extension (review comments)

daschuer commented on src/test/soundsourceproviderregistrytest.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    static QStringList getAllRegisteredProviderNamesForFileExtension(
+            const SoundSourceProviderRegistry& cut, QString fileExt) {
```
What stands cut for?

uklotzde replied:
"class under test" or "component under test", a common naming convention for 
unit tests:

http://c2.com/cgi/wiki?ClassUnderTest

On 11/25/2015 11:24 AM, Daniel Schürmann wrote:
>
> In src/test/soundsourceproviderregistrytest.cpp 
> <https://github.com/mixxxdj/mixxx/pull/605#discussion_r45850562>:
>
> > +            SoundSourceProviderPriority priorityHint = SoundSourceProviderPriority::DEFAULT) {
> > +        return SoundSourceProviderPointer(
> > +                new TestSoundSourceProvider(
> > +                        name, supportedFileExtensions, priorityHint));
> > +    }
> > +
> > +    SoundSourceProviderPointer createProvider(
> > +            QString name,
> > +            SoundSourceProviderPriority priorityHint = SoundSourceProviderPriority::DEFAULT) {
> > +        return SoundSourceProviderPointer(
> > +                new TestSoundSourceProvider(
> > +                        name, m_supportedFileExtensions, priorityHint));
> > +    }
> > +
> > +    static QStringList getAllRegisteredProviderNamesForFileExtension(
> > +            const SoundSourceProviderRegistry& cut, QString fileExt) {
>
> What stands cut for?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/605/files#r45850562>.
>

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,193 @@
+    if (m_registry.end() != i) {
+        return i.value();
+    } else {
+        return EMPTY_REGISTRATION_LIST;
```
return QList<SoundSourceProviderRegistration>()
would be unambiguous here.

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,193 @@
-            m_supportedFileNameRegex);
+void SoundSourceProviderRegistry::deregisterPluginLibrary(
+        const SoundSourcePluginLibraryPointer& pPluginLibrary) {
+    auto mapIter(m_registry.begin());
```
mapIter could be "registryIter"

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,193 @@
+        const QString& fileExtension,
+        const SoundSourceProviderPointer& pProvider) {
+    auto mapIter(m_registry.find(fileExtension));
+    if (m_registry.end() != mapIter) {
```
This loop code is duplicated below, could we move it to a helper function?

uklotzde replied:
Um, no, not exactly: One loop looks for the given provider and the other one 
for the given plugin library.

On 11/25/2015 01:46 PM, Daniel Schürmann wrote:
>
> In src/sources/soundsourceproviderregistry.cpp 
> <https://github.com/mixxxdj/mixxx/pull/605#discussion_r45862401>:
>
> >      }
> > -    foreach (const QString& supportedFileExtension, supportedFileExtensions) {
> > -        m_entries.insert(supportedFileExtension, entry);
> > +}
> > +
> > +void SoundSourceProviderRegistry::deregisterProviderForFileExtension(
> > +        const QString& fileExtension,
> > +        const SoundSourceProviderPointer& pProvider) {
> > +    auto mapIter(m_registry.find(fileExtension));
> > +    if (m_registry.end() != mapIter) {
>
> This loop code is duplicated below, could we move it to a helper function?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/605/files#r45862401>.
>

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,193 @@
+            registrationsForFileExtension.begin());
+    // Perform a linear search through the list
+    while (registrationsForFileExtension.end() != listIter) {
+        if (listIter->getProvider() == pProvider) {
```
I do not understand why there is no else branch and how it should work.  
Here the provider was already registered.  If the priority has changed the provider is removed and registered again. 
The priority remains the old one.

uklotzde replied:
I will delete this function since it is unused, untested, and the desired 
functionality is not clearly specified ;)

On 11/25/2015 01:54 PM, Daniel Schürmann wrote:
>
> In src/sources/soundsourceproviderregistry.cpp 
> <https://github.com/mixxxdj/mixxx/pull/605#discussion_r45863056>:
>
> > +    QList<SoundSourceProviderRegistration>::iterator listIter(
> > +            registrationsForFileExtension.begin());
> > +    insertRegistration(&registrationsForFileExtension, registration);
> > +}
> > +
> > +void SoundSourceProviderRegistry::reregisterProviderForFileExtension(
> > +        const QString& fileExtension,
> > +        const SoundSourceProviderPointer& pProvider,
> > +        SoundSourceProviderPriority providerPriority) {
> > +    QList<SoundSourceProviderRegistration>& registrationsForFileExtension =
> > +            m_registry[fileExtension];
> > +    QList<SoundSourceProviderRegistration>::iterator listIter(
> > +            registrationsForFileExtension.begin());
> > +    // Perform a linear search through the list
> > +    while (registrationsForFileExtension.end() != listIter) {
> > +        if (listIter->getProvider() == pProvider) {
>
> I do not understand why there is no else branch and how it should work.
>
> Here the provider was already registered. If the priority has changed the 
> provider is removed and registered again.
> The priority remains the old one.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/605/files#r45863056>.
>

daschuer commented on src/sources/soundsourceproviderregistry.cpp:
```diff
@@ -1,56 +1,193 @@
+    DEBUG_ASSERT(registration.getProvider());
+    QList<SoundSourceProviderRegistration>& registrationsForFileExtension =
+            m_registry[fileExtension];
+    QList<SoundSourceProviderRegistration>::iterator listIter(
```
unused listIter

daschuer commented on src/sources/soundsourceproviderregistry.h:
```diff
@@ -4,54 +4,109 @@
+    // Reregister a provider that is already registered with
+    // a new priority. It doesn't matter if the provider is
+    // built-in or provided by a plugin library.
+    void reregisterProviderForFileExtension(
```
This function seams to be unused. 
It should be called like: changeProviderPriorityForFileExtension()

daschuer commented on src/sources/soundsourceproviderregistry.h:
```diff
@@ -4,54 +4,109 @@
 public:
-    SoundSourceProviderPointer registerProvider(
+    // Registers a provider for all supported file extensions
+    // with their implicit priority hint.
```
with their (cooperative) priorityHind()

daschuer commented:
Cool! I think we can go this way. 

I am wondering when will the right date for a merge? With this PR, ffmpeg silently sneaks into Mixxx. 
This would be great for the variety of files we can play, but it will also likely introduce other issues by that. 

Since we are planing to release 2.1 short after 2.0 there might be not enough testing time. 
But on the other hand testing time does not start to run, if we do not merge it to master. 

Any thought? 
@illuusio @rryan @ywwg

illuusio commented:
@daschuer  I'm all in FFmpeg sneaking into 2.1 but let's not make it official yet just under the hood or something and on 2.2 make it more public.

uklotzde committed:
Delete unnecessary constant EMPTY_REGISTRATION_LIST (review comments)

uklotzde committed:
Renaming: mapIter -> registryIter (review comments)

uklotzde committed:
Delete unused function reregisterProviderForFileExtension() (review comments)

It's unclear how this function should behave if
- the priority remains unchanged
- the provider is not registered for the given file extension

uklotzde committed:
Delete unused listIter (review comments)

uklotzde committed:
Use the term "cooperative" for priority hint (review comments)

uklotzde committed:
Merge branch 'master' into SoundSourceProviderPriority

uklotzde committed:
Merge branch 'master' into SoundSourceProviderPriority

uklotzde committed:
Merge branch 'master' into SoundSourceProviderPriority

rryan commented:
> Any thought? 



> Cool! I think we can go this way.
> 
> I am wondering when will the right date for a merge? With this PR, ffmpeg silently sneaks into Mixxx. 
> This would be great for the variety of files we can play, but it will also likely introduce other issues by that.
> 
> Since we are planing to release 2.1 short after 2.0 there might be not enough testing time. 
> But on the other hand testing time does not start to run, if we do not merge it to master.
> 
> Any thought? 
> @illuusio @rryan @ywwg

Merging this PR wouldn't affect anything if the build server doesn't enable FFMPEG, right? If so, then merging it for 2.1 should be ok.

daschuer commented:
LGTM!
This means we can merge this. Correct @uklotzde?

uklotzde committed:
Merge branch 'master' into SoundSourceProviderPriority

uklotzde commented:
@daschuer No conflicts, ready for merging.

daschuer commented:
Cool, thank you.
